### PR TITLE
Fix lint warning

### PR DESCRIPTION
**Pull request by bmarty**
## Type of change

- [ ] Feature
- [ ] Bugfix
- [x] Technical
- [ ] Other :

## Content

I wanted to fix some lint warning, but end up doing more cleanup than expected.

@jmartinesp I have removed AndroidVersionTestOverrider in 7a0554c47d1287d156c1bf3c9e6af132e9177bc1. Let me know what you think!

## Motivation and context

Fix some lint warning and ensure the code follow some architecture pattern.

## Tests

- Smoke test, especially on language change.

## Tested devices

- [ ] Physical
- [ ] Emulator
- OS version(s):

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: vector/src/main/java/im/vector/app/core/ui/views/KeysBackupBanner.kt [hunk 1]
```diff
@@ -191,61 +163,12 @@ class KeysBackupBanner @JvmOverloads constructor(
...
-        object Hidden : State()
-
-        // Keys backup is not setup, numberOfKeys is the number of locally stored keys
-        data class Setup(val numberOfKeys: Int) : State()
-
-        // Keys backup can be recovered, with version from the server
-        data class Recover(val version: String) : State()
-
-        // Keys backup can be updated
-        data class Update(val version: String) : State()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/ui/views/KeysBackupBanner.kt:212]: This has been moved to the class `BannerState` in the ViewModel.

@BillCarsonFr this is worth noting that `Recover` and `Update` are actually never used.

File: vector/src/main/java/im/vector/app/features/workers/signout/ServerBackupStatusViewModel.kt [hunk 2]
```diff
@@ -51,29 +53,55 @@ data class ServerBackupStatusViewState(
...
-    object Hidden : BannerState()
+    // View will be Gone
+    object Hidden : BannerState
 
     // Keys backup is not setup, numberOfKeys is the number of locally stored keys
-    data class Setup(val numberOfKeys: Int) : BannerState()
+    data class Setup(val numberOfKeys: Int, val doNotShowAgain: Boolean) : BannerState
+
+    // Keys backup can be recovered, with version from the server
+    data class Recover(val version: String, val doNotShowForVersion: String) : BannerState
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/workers/signout/ServerBackupStatusViewModel.kt:67]: never used

File: vector/src/main/java/im/vector/app/features/workers/signout/ServerBackupStatusViewModel.kt [hunk 3]
```diff
@@ -51,29 +53,55 @@ data class ServerBackupStatusViewState(
...
+
+    // Keys backup can be updated
+    data class Update(val version: String, val doNotShowForVersion: String) : BannerState
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/workers/signout/ServerBackupStatusViewModel.kt:70]: never used...

File: vector/src/main/java/im/vector/app/core/ui/views/KeysBackupBanner.kt [hunk 4]
```diff
@@ -191,61 +163,12 @@ class KeysBackupBanner @JvmOverloads constructor(
...
-         */
-        private const val BANNER_UPDATE_DO_NOT_SHOW_FOR_VERSION = "BANNER_UPDATE_DO_NOT_SHOW_FOR_VERSION"
-
-        /**
-         * Inform the banner that a Recover has been done for this version, so do not show the Recover banner for this version.
-         */
-        fun onRecoverDoneForVersion(context: Context, version: String) {
-            DefaultSharedPreferences.getInstance(context).edit {
-                putString(BANNER_RECOVER_DO_NOT_SHOW_FOR_VERSION, version)
-            }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/ui/views/KeysBackupBanner.kt:248]: moved to the ViewModel

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/util/BuildVersionSdkIntProvider.kt [hunk 5]
```diff
@@ -26,6 +28,7 @@ interface BuildVersionSdkIntProvider {
      * Checks the if the current OS version is equal or greater than [version].
      * @return A `non-null` result if true, `null` otherwise.
      */
+    @ChecksSdkIntAtLeast(parameter = 0, lambda = 1)
```
Comment (ouchadam) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/util/BuildVersionSdkIntProvider.kt:31]: TIL :100:

Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/util/BuildVersionSdkIntProvider.kt:31]: This was also new to me, I learn about it looking at the lint warning. Very useful!

File: vector/src/main/java/im/vector/app/core/utils/RingtoneUtils.kt [hunk 6]
```diff
@@ -17,103 +17,109 @@
...
- *
- * @return the [Uri] of the currently set [Ringtone]
- * @see Ringtone
- */
-fun getCallRingtoneUri(context: Context): Uri? {
-    val callRingtone: String? = DefaultSharedPreferences.getInstance(context)
-            .getString(VectorPreferences.SETTINGS_CALL_RINGTONE_URI_PREFERENCE_KEY, null)
+        callRingtone?.let {
+            return Uri.parse(it)
+        }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/utils/RingtoneUtils.kt:54]: Note to reviewers: better to compare ignoring space.

File: vector/src/main/java/im/vector/app/features/voice/VoiceRecorderProvider.kt [hunk 7]
```diff
@@ -37,8 +40,11 @@ class VoiceRecorderProvider @Inject constructor(
         }
     }
 
+    @ChecksSdkIntAtLeast(api = 29)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/voice/VoiceRecorderProvider.kt:43]: Actually this is the opposite here, let me double check

File: vector/src/main/java/im/vector/app/features/workers/signout/ServerBackupStatusAction.kt [hunk 8]
```diff
@@ -14,18 +14,12 @@
...
 
-package im.vector.app.core.di
+package im.vector.app.features.workers.signout
 
-import android.content.Context
-import android.content.SharedPreferences
-import androidx.preference.PreferenceManager
+import im.vector.app.core.platform.VectorViewModelAction
 
-object DefaultSharedPreferences {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/workers/signout/ServerBackupStatusAction.kt:23]: This is removed

File: vector/src/main/java/im/vector/app/features/voice/VoiceRecorderProvider.kt (hunk 7 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/voice/VoiceRecorderProvider.kt]: fa5223cf36c1132dbeca1d2024cd4cddb3173639

**Comment by jmartinesp** [comment]
Removing `AndroidVersionTestOverrider` is definitely an option. A couple of weeks ago I also proposed using some test rule to avoid having to create and inject a component to override SDK versions when we only need those for testing. Something like:

```kotlin
@get:Rule
val rule = AndroidVersionTestRule()

@Test
@TestSdkVersion(Build.VERSION_CODES.Q)
fun provideVoiceRecorderOnAndroidQReturnsQRecorder() {
    provider.provideVoiceRecorder().shouldBeInstanceOf(VoiceRecorderQ::class)
}
```

This rule would effectively do the same as the removed overrider, but it would only be applied to test cases with the `@TestSdkVersion` annotation and it would automatically be cleared after the test case is finished, I tried a simple implementation and it worked out of the box. 

It seems a lot cleaner to me to keep something that's only useful for testing purposes at the test layer, but that's just my opinion. Also, this approach doesn't need us to add the `@SuppressLint("NewApi")` everywhere, although maybe that would be fixed by the `@ChecksSdkIntAtLeast(api = X)` annotation that you added here?

**Comment by bmarty** [comment]
Yes, using `@ChecksSdkIntAtLeast` allow to remove `@SuppressLint("NewApi")`, but it's not perfect, see https://github.com/vector-im/element-android/pull/7159/commits/fa5223cf36c1132dbeca1d2024cd4cddb3173639 (compilation OK before and after the change)

I will add a commit to remove more `@SuppressLint("NewApi")`

On your remark about test, I agree that everything related to the test should stay at the test layer. Sadly I cannot find it anymore, but we has a test which tested a system upgrade. So that we add several set of the SDK level during the test. Using `@TestSdkVersion` could not work in this case, IIUC.

**Comment by jmartinesp** [comment]
Sadly I cannot find it anymore, but we has a test which tested a system upgrade. So that we add several set of the SDK level during the test. Using `@TestSdkVersion` could not work in this case, IIUC.
> 
> **EDIT** The test I was looking for is
> 
> https://github.com/vector-im/element-android/blob/b5aedd4626af6593fd0f737e393e573831c82356/matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/api/securestorage/SecretStoringUtilsTest.kt#L176

We could add a method to that rule to manually set the SDK version if needed, even something like:

```
versionRule.using(Build.VERSION_CODES.Q) { 
    // Execute the code with the overridden version here
   ...
}
// After it, version is restored to normal
```

**Comment by bmarty** [comment]
Some tests are failing, will handle that

```
im.vector.app.features.pin.lockscreen.crypto.migrations.LegacyPinCodeMigratorTests > migratePinCodeL[test(AVD) - 9] FAILED 
	io.mockk.MockKException: no answer found for: BuildVersionSdkIntProvider(#84).isAtLeast(23)
	at io.mockk.impl.stub.MockKStub.defaultAnswer(MockKStub.kt:93)

im.vector.app.features.pin.lockscreen.crypto.migrations.LegacyPinCodeMigratorTests > migratePinCodeM[test(AVD) - 9] FAILED 
	io.mockk.MockKException: no answer found for: BuildVersionSdkIntProvider(#89).isAtLeast(23)
	at io.mockk.impl.stub.MockKStub.defaultAnswer(MockKStub.kt:93)
```

**Comment by bmarty** [comment]
(I will rebase this branch)

**Review by ouchadam — APPROVED** [review]

**Comment by bmarty** [comment]
OK, rebased, if CI is happy, I guess we can merge it.

**Pull request merged**